File: src/main/java/de/maggiwuerze/xdccloader/events/EventPublisher.java
package de.maggiwuerze.xdccloader.events;

import de.maggiwuerze.xdccloader.events.download.DownloadUpdateEvent;
import de.maggiwuerze.xdccloader.irc.IrcBot;
import de.maggiwuerze.xdccloader.model.download.Download;
import de.maggiwuerze.xdccloader.model.download.DownloadState;
import de.maggiwuerze.xdccloader.service.DownloadService;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.context.ApplicationEventPublisher;
import org.springframework.messaging.simp.SimpMessagingTemplate;
import org.springframework.stereotype.Component;

@Component
@RequiredArgsConstructor
@Slf4j
public class EventPublisher {

	private static final String MESSAGE_PREFIX = "/topic";
	private final DownloadService downloadService;
	private final SimpMessagingTemplate websocket;
	private final ApplicationEventPublisher applicationEventPublisher;

	public void updateDownloadState(DownloadState state, Download download) {

		download.setStatus(state);
		downloadService.update(download);

		DownloadUpdateEvent downloadUpdateEvent = new DownloadUpdateEvent(this, download.getId());
		applicationEventPublisher.publishEvent(downloadUpdateEvent);
	}

	public void sendWebsocketEvent(SocketEvents event, Object payload) {

		this.websocket.convertAndSend(
			MESSAGE_PREFIX + event.getRoute(), payload);
	}

	public void handleError(IrcBot bot, Exception exception) {

		bot.stopBotReconnect();
		Download download = downloadService.getById(bot.getDownloadId());
		String message = String.format(DownloadState.ERROR.getExternalString(), exception.getMessage());
		try {
			download.setStatusMessage(message);
			updateDownloadState(DownloadState.ERROR, download);
		} catch (Exception e) {
			log.error(e.getMessage());
		}
	}
}

File: src/main/kotlin/de/maggiwuerze/xdccwebloader/events/EventPublisher.kt
package de.maggiwuerze.xdccwebloader.events

import de.maggiwuerze.xdccwebloader.events.download.DownloadUpdateEvent
import de.maggiwuerze.xdccwebloader.irc.IrcBot
import de.maggiwuerze.xdccwebloader.model.download.Download
import de.maggiwuerze.xdccwebloader.model.download.DownloadState
import de.maggiwuerze.xdccwebloader.service.DownloadService
import org.slf4j.LoggerFactory
import org.springframework.context.ApplicationEventPublisher
import org.springframework.messaging.simp.SimpMessagingTemplate
import org.springframework.stereotype.Component

@Component
class EventPublisher(
    private val downloadService: DownloadService,
    private val websocket: SimpMessagingTemplate,
    private val applicationEventPublisher: ApplicationEventPublisher
) {

    val log = LoggerFactory.getLogger(this.javaClass.name)

    fun updateDownloadState(state: DownloadState, download: Download) {
        download.status = state
        downloadService.update(download)

        DownloadUpdateEvent(this, download.id).also { downloadUpdateEvent ->
            applicationEventPublisher.publishEvent(downloadUpdateEvent)
        }
    }

    fun sendWebsocketEvent(event: SocketEvents, payload: Any) {
        this.websocket.convertAndSend(
            MESSAGE_PREFIX + event.route, payload
        )
    }

    fun handleError(bot: IrcBot, exception: Exception) {
        bot.stopBotReconnect()
        bot.sendIRC().quitServer()
        downloadService.getOrThrow(bot.downloadId).let { download ->

            try {
                download.statusMessage = exception.message ?: "Unknown error"
                updateDownloadState(DownloadState.ERROR, download)
            } catch (e: Exception) {
                log.error(e.message)
            }

        }
    }

    companion object {
        private const val MESSAGE_PREFIX = "/topic"
    }
}
